Тетрадка предназначена для студентов образвательного проекта Booster MDS использование вне проекта запрещено.

##### Задания не несут смысловой нагрузки для реальной жизни, они сделаны лишь для того, чтоб научить использовать Pandas

#### Каждое задание оценивает в 1 балл, для успешной сдачи задания нужно набрать 12 баллов

This database contains 76 attributes, but all published experiments refer to using a subset of 14 of them. In particular, the Cleveland database is the only one that has been used by ML researchers to this date. The "goal" field refers to the presence of heart disease in the patient. It is integer valued from 0 (no presence) to 1.

https://www.kaggle.com/ronitf/heart-disease-uci

___

In [1]:
import pandas as pd
import numpy as np

In [2]:
data = pd.read_csv('heart_pandas.csv')

In [3]:
data.head(5)

age     sex   chest_pain_type  resting_blood_pressure  cholesterol  \
0   63    male  non-anginal pain                     145          233   
1   37    male   atypical angina                     130          250   
2   41  female    typical angina                     130          204   
3   56    male    typical angina                     120          236   
4   57  female    typical angina                     120          354   

     fasting_blood_sugar               rest_ecg  max_heart_rate_achieved  \
0  greater than 120mg/ml                 normal                      150   
1    lower than 120mg/ml  ST-T wave abnormality                      187   
2    lower than 120mg/ml                 normal                      172   
3    lower than 120mg/ml  ST-T wave abnormality                      178   
4    lower than 120mg/ml  ST-T wave abnormality                      163   

  exercise_induced_angina  st_depression   st_slope  num_major_vessels  \
0                      no            2.3  upsloping                  0   
1                      no            3.5  upsloping                  0   
2                      no            1.4       flat                  0   
3                      no            0.8       flat                  0   
4                     yes            0.6       flat                  0   

    thalassemia  target  
0        normal       1  
1  fixed defect       1  
2  fixed defect       1  
3  fixed defect       1  
4  fixed defect       1

#### Описание признаков

**age**: The person's age in years

**sex**: The person's sex (1 = male, 0 = female)

**cp**: The chest pain experienced (Value 1: typical angina, Value 2: atypical angina, Value 3: non-anginal pain, Value 4: asymptomatic)

**trestbps**: The person's resting blood pressure (mm Hg on admission to the hospital)

**chol**: The person's cholesterol measurement in mg/dl

**fbs**: The person's fasting blood sugar (> 120 mg/dl, 1 = true; 0 = false)

**restecg**: Resting electrocardiographic measurement (0 = normal, 1 = having ST-T wave abnormality, 2 = showing probable or definite left ventricular hypertrophy by Estes' criteria)

**thalach**: The person's maximum heart rate achieved

**exang**: Exercise induced angina (1 = yes; 0 = no)

**oldpeak**: ST depression induced by exercise relative to rest ('ST' relates to positions on the ECG plot. See more here)

**slope**: the slope of the peak exercise ST segment (Value 1: upsloping, Value 2: flat, Value 3: downsloping)

**ca**: The number of major vessels (0-3)

**thal**: A blood disorder called thalassemia (3 = normal; 6 = fixed defect; 7 = reversable defect)

**target**: Heart disease (0 = no, 1 = yes)

> 1. age 
> 2. sex 
> 3. chest pain type (4 values) 
> 4. resting blood pressure 
> 5. serum cholestoral in mg/dl 
> 6. fasting blood sugar > 120 mg/dl
> 7. resting electrocardiographic results (values 0,1,2)
> 8. maximum heart rate achieved 
> 9. exercise induced angina 
> 10. oldpeak = ST depression induced by exercise relative to rest 
> 11. the slope of the peak exercise ST segment 
> 12. number of major vessels (0-3) colored by flourosopy 
> 13. thal: 3 = normal; 6 = fixed defect; 7 = reversable defect

___

**Сколько мужчин в датасете? Сколько женщин? (sex)**

In [4]:
## Ваш код
data['sex'].value_counts()

male      207
female     96
Name: sex, dtype: int64

In [5]:
## Ваш ответ:    Мужчин: 207
#                Женщин: 96

 **Какой процент мужчин в датасете? (решите в одну строчку, не используя предыдущий результат. Не забудте знак процента)**

In [6]:
## Ваш код
f"{data['sex'].eq('male').mean() * 100:.3f}%"

'68.317%'

In [ ]:
## Ваш ответ: 68.317%

**Сколько мужчин имеют заболевание сердца? Сколько женщин имеют заболевание сердца?**

In [7]:
## Ваш код
data[data['target'] == 1].sex.value_counts()

male      93
female    72
Name: sex, dtype: int64

In [8]:
## Ваш ответ:
#  Мужчин с ЗС: 93
#  Женщин с ЗС: 72

**Какую долю, от общего числа пациентов, занимают мужчины не имеющие болезнь сердца?**

In [9]:
## Ваш код
male_without_hd = data[data['target'] == 0].sex.value_counts()['male']

print(f'{male_without_hd / len(data) * 100:.2f}%')

37.62%


In [10]:
## Ваш ответ: 37.62%

**Сколько лет самому молодому пациенту, страдающему болезнью сердца?**

In [11]:
## Ваш код
data.sort_values(by=['target', 'age'], ascending=[False, True]).iloc[0]['age']

29

In [ ]:
## Ваш ответ: 29 лет.

**Сколько лет самому возрастному пациенту, у которого нет проблем с сердцем?**

In [12]:
## Ваш код
data.sort_values(by=['target', 'age'], ascending=[True, False]).iloc[0]['age']

77

In [13]:
## Ваш ответ: 77 лет.

**Сколько лет самой молодой женщине, которая страдает болезнью сердца?**

In [14]:
## Ваш код
data[(data['target'] == 1) & (data['sex'] == 'female')].describe()['age']['min']

34.0

In [22]:
## Ваш ответ: 34 года

**Какой средний возраст женщин?**

In [15]:
## Ваш код
data[data['sex'] == 'female']['age'].mean()

55.677083333333336

In [24]:
## Ваш ответ: 55.6

**Каковы средние значения и среднеквадратичные отклонения возраста тех, кто страдают болезнью сердца?**

In [16]:
## Ваш код
patients_age = data[data['target'] == 1]['age']
print(f'mean: {patients_age.mean()}\n'
      f'std:  {patients_age.std()}')

mean: 52.4969696969697
std:  9.550650751946778


In [25]:
## Ваш ответ:     mean: 52.4969696969697
#                 std:  9.550650751946778

**Правда ли, что люди не болеющие болезнью сердца имеют уровня холестерина меньше среднего? (chol)**

In [17]:
## Ваш код
bool(data[data['target'] == 0]['cholesterol'].mean() < data['cholesterol'].mean())

False

In [18]:
## Ваш ответ:    неправда

**Выведите статистику rest_ecg для всех числовых признаков, его максимально и среднее значение (используйте groupby(), решите в одну строчку)**

In [19]:
## Ваш код
data.groupby('rest_ecg').agg([np.max, np.mean])

age            resting_blood_pressure  \
                             amax       mean                   amax   
rest_ecg                                                              
ST-T wave abnormality          71  52.914474                    180   
left ventricular hypertrophy   76  61.000000                    180   
normal                         77  55.687075                    200   

                                         cholesterol              \
                                    mean        amax        mean   
rest_ecg                                                           
ST-T wave abnormality         129.065789         354  237.269737   
left ventricular hypertrophy  140.500000         327  261.750000   
normal                        134.027211         564  255.142857   

                             max_heart_rate_achieved              \
                                                amax        mean   
rest_ecg                                                           
ST-T wave abnormality                            194  151.960526   
left ventricular hypertrophy                     140  125.750000   
normal                                           202  147.904762   

                             st_depression           num_major_vessels  \
                                      amax      mean              amax   
rest_ecg                                                                 
ST-T wave abnormality                  5.6  0.879605                 4   
left ventricular hypertrophy           4.4  2.725000                 3   
normal                                 6.2  1.159184                 4   

                                       target            
                                  mean   amax      mean  
rest_ecg                                                 
ST-T wave abnormality         0.638158      1  0.631579  
left ventricular hypertrophy  1.000000      1  0.250000  
normal                        0.816327      1  0.462585

In [35]:
## Ваш ответ:    ответ - результат выполнения кода выше.

**Посчитайте у кого уровень депрессии при физический нагрузке выше (в среднем), среди мужчин страдающих болезнью сердца или среди женщин не страдающих болезнью сердца (st_depression)**

In [20]:
## Ваш код
data.pivot_table(['st_depression'], 
                 ['target', 'sex'],
                 aggfunc='mean')

st_depression
target sex                  
0      female       1.841667
       male         1.531579
1      female       0.554167
       male         0.605376

In [37]:
## Ваш ответ: уровень депрессии сегмента ST выше у женщин,
#             не имеющих сердечных заболеваний,
#             что довольно странно...или данные были интерпретированны неверно

**Посчтитайте максимальный и минимальный уровень холестерина для каждого типа chest_pain_type, rest_ecg, thalassemia. Пишите код оптимально, можно использовать циклы**

In [21]:
## Ваш код
for feature in 'chest_pain_type', 'rest_ecg', 'thalassemia':
    print(data.pivot_table(['cholesterol'],
                           [feature],
                           aggfunc=['max', 'min']))

                         max         min
                 cholesterol cholesterol
chest_pain_type                         
atypical angina          564         126
non-anginal pain         298         182
typical angina           409         131
                                     max         min
                             cholesterol cholesterol
rest_ecg                                            
ST-T wave abnormality                354         126
left ventricular hypertrophy         327         197
normal                               564         149
                          max         min
                  cholesterol cholesterol
thalassemia                              
fixed defect              417         141
normal                    318         169
reversable defect         564         126


In [ ]:
## Ваш ответ:
#                         max         min
#                         chol        chol
# chest_pain_type                         
# atypical angina         564         126
# non-anginal pain        298         182
# typical angina          409         131
#
#                         max         min
#                         chol        chol
# rest_ecg                                            
# ST-T  abnormality       354         126
# left-v hypertrophy      327         197
# normal                  564         149
#
#                         max         min
#                         chol        chol
# thalassemia                              
# fixed defect            417         141
# normal                  318         169
# reversable defect       564         126

**Сколько значений может принимать каждый из категориальных признаков?**

In [22]:
## Ваш код
for feature in data.describe(include=[np.bool, np.object]):
    print(f'{feature}: {len(pd.unique(data[feature]))}')

sex: 2
chest_pain_type: 3
fasting_blood_sugar: 2
rest_ecg: 3
exercise_induced_angina: 2
st_slope: 2
thalassemia: 3


In [ ]:
## Ваш ответ:

#     sex: 2
#     chest_pain_type: 3
#     fasting_blood_sugar: 2
#     rest_ecg: 3
#     exercise_induced_angina: 2
#     st_slope: 2
#     thalassemia: 3
# (?) target: 2

**У какого категориального признака наблюдается самый сильный диcбаланс классов?**

In [23]:
## Ваш код
df = pd.DataFrame(columns=['feature','imbalance', 'max', 'min'])

for feature in data:
    feat_max = data[feature].value_counts().max()
    feat_min = data[feature].value_counts().min()
    imbalance = feat_min/feat_max * 100
    # append row to df
    df = df.append({'feature': feature,
                    'imbalance' : imbalance,
                    'max' : feat_max,
                    'min' : feat_min} , ignore_index=True)

In [24]:
df.sort_values(by='imbalance').reset_index(drop=True)

feature  imbalance  max  min
0             st_depression   1.010101   99    1
1                  rest_ecg   2.631579  152    4
2    resting_blood_pressure   2.702703   37    1
3         num_major_vessels   2.857143  175    5
4                       age   5.263158   19    1
5   max_heart_rate_achieved   9.090909   11    1
6           chest_pain_type  11.917098  193   23
7               thalassemia  12.048193  166   20
8               cholesterol  16.666667    6    1
9       fasting_blood_sugar  17.441860  258   45
10                      sex  46.376812  207   96
11  exercise_induced_angina  48.529412  204   99
12                   target  83.636364  165  138
13                 st_slope  88.198758  161  142

In [ ]:
## Ваш ответ: самый сильный диcбаланс классов у признака st_depression

---

@Booster
Мамкин Data Scientist

tg @mommyscience